### PR TITLE
BufModifiedSet event

### PR DESCRIPTION
**Pull request by romgrk**
POC implementation, comments welcome on the viability of the PR.

TBD:
 - As per comment in `get_option_value()`: `// Special case: 'modified' is b_changed, but we also want to consider it set when 'ff' or 'fenc' changed.`. We might want to extend the logic here to fulfill that behavior.
 - Event when non-current buffer is modified

### CODE REVIEW

**Comment by bfredl** [comment]
Not sure of the name `BufModified`, it would naïvely imply to me only being trigged on modified, but this is also for nomodified, Maybe `BufModifiedSet` in analogy with `OptionSet` (which has precendent in being triggered also on _unsetting_ options) ?

**Comment by bfredl** [comment]
What happens if non-current buffer is modified? I think the autocmd should then trigger in the context of that buffer (changing a non-current buffer does a `aucmd_prepbuf` anyway)

**Comment by romgrk** [comment]
Yes, it should also be triggered in those cases. While doing this first implementation I just searched for places where `b_changed` was modified. The other moments where this should be triggered would be:
 - buffer-modifying functions: `append(bufline)`, `set(buf)line`, `nvim_buf_set_lines`
 - `ff` or `ffenc` modified
 - any other place that you can think of?

Also if you have time would it be possible to review #13117 first? Unless you think the implementation there isn't sound and this has more chances to be emrged first, I'd prefer to close that first because it's touching the same areas. I was also wondering if it wouldn't be benecifial to refactor a bit some of the autocmd triggering code because I a few repeated blocks of logic.

**Comment by bfredl** [comment]
I hope to unrepeat some repeated code when i finally do set_text

**Comment by romgrk** [comment]
@bfredl would we want to merge this one as well before I begin work on refactoring/fixing the autocommands? It has the exact same structure & logic as the other atocommands we discussed: textchanged, cursormoved, windowscrolled. I've added some test just now to test the current-buffer-modified functionality, and non-current-modified would be implemented with the others.

**Comment by romgrk** [comment]
Test failing because neovim is executing the test in `main`'s `exe_commands` but the checking code runs only in the event loop, therfore it says modified once even though the things was modified twice. I'll check for a way to delay the test after entering the event loop but first my functional tests don't build in local -_- Ima try to fix that first

**Comment by romgrk** [comment]
ping @bfredl sorry for the pressure, see above comments, if this is merged before #13247 I'd rather do it soon to avoid conflicts.

File: src/nvim/buffer.c [hunk 1]
```diff
@@ -97,6 +97,13 @@ typedef enum {
   kBffInitChangedtick = 2,
 } BufFreeFlags;
 
+/// Allocate new buf structure and initialize it.
+buf_T *buf_alloc(void) {
+  buf_T *buf = xcalloc(1, sizeof(buf_T));
+  buf->b_changed_notified = true;
```
Comment (bfredl) [inline src/nvim/buffer.c:103]: This should just be below `buf->b_signcols_max = -1;` in `buflist_new` ( a bit weirdly named function but it is already the only entry point for allocating buffers.)

File: src/nvim/change.c [hunk 2]
```diff
@@ -129,6 +129,7 @@ void changed(void)
 void changed_internal(void)
 {
   curbuf->b_changed = true;
+  curbuf->b_changed_notified = false;
```
Comment (bfredl) [inline src/nvim/change.c:132]: Usually we prefer invalidiation flags to be the other way around, i e a flag that is `true` if a later action is needed.

File: test/functional/autocmd/bufmodifiedset_spec.lua [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+describe('BufModified', function()
+  before_each(clear)
+
+  it('is triggered when modified and un-modified', function()
+    source([[
+    let g:modified = 0
+    autocmd BufModifiedSet * let g:modified += 1
+    ]])
+    request("nvim_command", [[normal! aa\<Esc>]])
```
Comment (bfredl) [inline test/functional/autocmd/bufmodifiedset_spec.lua:17]: I  would add a `eq(1, eval('g:modified'))` in between just to be extra paranoid :)

**Comment by bfredl** [comment]
@romgrk LGTM, just see a few comments.

**Comment by romgrk** [comment]
Done, and Travis says okay.

**Comment by justinmk** [comment]
What actually is the use-case for this? E.g. vs "buffer updates" (`nvim_buf_attach`) or a `TextChanged` handler?

**Comment by romgrk** [comment]
I use it in barbar.nvim (tabline plugin) and in some other custom local plugins. I guess those other methods would work, it just seems a bit wasteful considering they might be called on every key press (right?). In general moving plugin code away from anything that can affect key press latency sounds like a good thing to me

**Comment by justinmk** [comment]
Thanks. Different question, why not make `OptionSet modified` work instead of adding a new event?

All of this should have been called out in the PR or a linked issue...

**Comment by romgrk** [comment]
Great question, maybe me-from-5-years-ago had reasons but I don't know them.

**Pull request merged**